utils/s3/client.cc

```diff
@@ -1092,6 +1093,117 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
     return data_sink(std::make_unique<upload_jumbo_sink>(shared_from_this(), std::move(object_name), max_parts_per_piece, as));
 }
 
+class client::chunked_download_source final : public seastar::data_source_impl {
+    shared_ptr<client> _client;
+    sstring _object_name;
+    seastar::abort_source* _as;
+    range _range;
+    static constexpr size_t _max_buffers_size = 5_MiB;
+    std::deque<temporary_buffer<char>> _buffers;
+    size_t _buffers_size = 0;
+    semaphore _buffers_sem;
+    bool _is_closing = false;
+    bool _is_done = false;
+    seastar::condition_variable _buffer_cv;
+    future<> _filling_fiber = make_ready_future<>();
+
+    future<> make_filling_fiber() {
+        while (true) {
+            try {
+                s3l.trace("Fiber starts cycle for object '{}'", _object_name);
+                if ((_as && _as->abort_requested()) || _is_closing) {
+                    s3l.trace("Abandoning fiber for object '{}'", _object_name);
+                    co_return;
+                }
+                auto download_source = _client->make_download_source(_object_name, _range, _as);
+                s3l.trace("Fiber for object '{}' will do work? {}", _object_name, (_buffers_size < _max_buffers_size && !_is_closing) ? "yes" : "no");
+                while (_buffers_size < _max_buffers_size && !_is_closing) {
+                    s3l.trace("Fiber for object '{}' will try to download within range {}-{}", _object_name, _range.off, _range.len);
+                    auto buf = co_await download_source.get();
+                    s3l.trace("Fiber for object '{}' received buffer of {} bytes", _object_name, buf.size());
+                    _range.off += buf.size();
+                    _range.len -= buf.size();
+                    _buffers_size += buf.size();
+                    if (buf.empty()) {
+                        _is_done = true;
+                        _buffer_cv.broadcast();
+                        co_await download_source.close();
+                        s3l.trace("Fiber for object '{}' signals EOS", _object_name);
+                        co_return;
+                    }
+                    {
+                        auto units = co_await get_units(_buffers_sem, 1);
+                        s3l.trace("Fiber for object '{}' pushes {} bytes buffer", _object_name, buf.size());
+                        _buffers.emplace_back(std::move(buf));
```

Comment:
A buffer sits in RAM and consumes memory until .get() withdraws it. This pinned memory must be accounted into client::_memory semaphore with claim_memory()